src/tabmat/constructor.py

```diff
@@ -133,64 +131,177 @@ def from_pandas(
                     mxcolidx += cat.shape[1]
                 elif cat_position == "end":
                     indices.append(np.arange(cat.shape[1]))
-        # All other numerical dtypes (needs to be after pd.SparseDtype)
-        elif is_numeric_dtype(coldata):
-            # check if we want to store as sparse
+        elif pd.api.types.is_numeric_dtype(coldata):
             if (coldata != 0).mean() <= sparse_threshold:
-                if not isinstance(coldata.dtype, pd.SparseDtype):
-                    fill_value = False if is_bool_dtype(coldata) else 0  # type: ignore
-                    sparse_dtype = pd.SparseDtype(coldata.dtype, fill_value=fill_value)
-                    sparse_dfcols.append(coldata.astype(sparse_dtype))
-                else:
-                    sparse_dfcols.append(coldata)
-                sparse_mxidx.append(mxcolidx)
+                sparse_columns.append(colname)
+                sparse_indices.append(mxcolidx)
                 mxcolidx += 1
             else:
-                dense_dfidx.append(dfcolidx)
-                dense_mxidx.append(mxcolidx)
+                dense_columns.append(colname)
+                dense_indices.append(mxcolidx)
                 mxcolidx += 1
 
-        # dtype not handled yet
         else:
-            ignored_cols.append((dfcolidx, colname))
+            ignored_cols.append(colname)
 
     if len(ignored_cols) > 0:
         warnings.warn(
             f"Columns {ignored_cols} were ignored. Make sure they have a valid dtype."
         )
-    if len(dense_dfidx) > 0:
-        matrices.append(
-            DenseMatrix(
-                df.iloc[:, dense_dfidx].astype(dtype),
-                column_names=df.columns[dense_dfidx],
-                term_names=df.columns[dense_dfidx],
-            )
-        )
-        indices.append(dense_mxidx)
+    if dense_columns:
+        matrices.append(_dense_matrix(df, dense_columns, dtype))
```

Comment:
The [problem in glum](https://github.com/Quantco/glum/actions/runs/9772131161/job/26980639512) comes from here. It should be df[dense_columns]. However, using column names doesn't work here because of potential duplicates. This is why we had a logic to use the column index.